Psychtoolbox/PsychHardware/EyelinkToolbox/EyelinkBasic/PsychEyelinkDispatchCallback.m

```diff
@@ -129,326 +161,419 @@
+        if el.winInfo.StereoMode ~= 0 
+            drawScreens = 2; % stereoscopic drawing
+        else
+            drawScreens = 1; % non-stereoscopic drawing
+        end
+        
+        for it = 0:drawScreens-1 
+            Screen('SelectStereoDrawBuffer', eyewin, it); % select left eye window
+            oldFont=Screen(eyewin,'TextFont',el.msgfont);
+            oldFontSize=Screen(eyewin,'TextSize',el.msgfontsize);
+            DrawFormattedText(eyewin, el.helptext, 20, 20, el.msgfontcolour, [], [], [], 1);
+            
+            if el.displayCalResults && ~isempty(msg)
+                DrawFormattedText(eyewin, msg, 20, 150, el.msgfontcolour, [], [], [], 1);
+            end
+            
+            Screen(eyewin,'TextFont',oldFont);
+            Screen(eyewin,'TextSize',oldFontSize);
+        end
+        
+        
     end
-    % imgtitle
-    % if title is provided, we also draw title
-    if ~isempty(eyelinktex) && exist( 'imgtitle', 'var') && ~isempty(imgtitle)
-        %oldFont=Screen(eyewin,'TextFont',el.imgtitlefont);
-        %oldFontSize=Screen('TextSize',eyewin,el.imgtitlefontsize);
-        rect=Screen('TextBounds', eyewin, imgtitle );
-        [w2, h2]=RectSize(rect);
-
-        % added by NJ as a quick way to prevent over drawing and to clear text
-        if newImage || isempty(lasttitle) || ~strcmp(imgtitle,lasttitle)
-            if -1 == Screen('WindowKind', offscreen)
-                Screen('Close', offscreen);
+
+
+
+    function  imgtitle=EyelinkDrawCameraImage(eyewin, el, eyelinktex, imgtitle, newImage)
+        persistent lasttitle;
+        %         global dh dw offscreen;
+        if el.winInfo.StereoMode ~= 0 
+            drawScreens = 2; % stereoscopic drawing
+        else
+            drawScreens = 1; % non-stereoscopic drawing
+        end
+        
+        for it = 0:drawScreens-1 
+            try
+                if ~isempty(eyelinktex)
+                    Screen('SelectStereoDrawBuffer', eyewin, it); % select left eye window
+                    eyerect=Screen('Rect', eyelinktex);
+                    % we could cash some of the below values....
+                    wrect=Screen('Rect', eyewin);
+                    [width, heigth]=Screen('WindowSize', eyewin);
+                    dw=round(el.eyeimgsize/100*width);
+                    dh=round(dw * eyerect(4)/eyerect(3));
+                    
+                    drect=[ 0 0 dw dh ];
+                    drect=CenterRect(drect, wrect);
+                    Screen('DrawTexture', eyewin, eyelinktex, [], drect);
+                end
+                % imgtitle
+                % if title is provided, we also draw title
+                if ~isempty(eyelinktex) && exist( 'imgtitle', 'var') && ~isempty(imgtitle)
+                    Screen('SelectStereoDrawBuffer', eyewin, it); % select left eye window
+                    rect=Screen('TextBounds', eyewin, imgtitle );
+                    [w2, h2]=RectSize(rect);
```

Comment:
This whole stuff with drawing into offscreen windows was an optimization of your colleague NJ, which is now broken/meaningless in this code, since adding stereo drawing and maybe changing the dontclear flag of Flip.

What one would want to do is:

Only when at first run or if the imgtitle has changed, one would (re-)create an offscreen window, set its proper text size and style, draw the text into it once. Then just draw the offscreen window via DrawTexture once for the left eye, once for the right eye, as part of the for loop. All other stuff outside the for loop.

Generally, text handling ie. Screen('TextBounds'), Screen('DrawText') and DrawFormattedText() are relatively slow/expensive functions one wants to avoid needless invocation when possible. However, I don't know if this really matters for PsychEyelinkDispatchCallback.m, given that the Gazetracker setup and calibration phases are probably not such timing sensitive operations in an experiment. So it'd be more about giving a good example to others.

Reply:
will rework the previous optimizations in a future revision